src/librustc_lint/types.rs

```diff
@@ -176,6 +177,23 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TypeLimits {
                             _ => bug!(),
                         };
                         if lit_val < min || lit_val > max {
+                            let parent_id = cx.tcx.hir.get_parent_node(e.id);
+                            if let hir_map::NodeExpr(parent_expr) = cx.tcx.hir.get(parent_id) {
+                                if let hir::ExprCast(..) = parent_expr.node {
+                                    if let ty::TyChar = cx.tables.expr_ty(parent_expr).sty {
+                                        let mut err = cx.struct_span_lint(
+                                                             OVERFLOWING_LITERALS,
+                                                             e.span,
+                                                             "only u8 can be casted into char");
+                                        err.span_suggestion(e.span,
```

Comment:
This should be a span enclosing all of `lit as char` (probably the `parent_expr`'s span). If you don't want to have overlapping spans, use the same span in the `DiagnosticBuilder` and for the suggestion.